oauth2_http/java/com/google/auth/mtls/SecureConnectProvider.java

```diff
@@ -0,0 +1,161 @@
+
+  private String metadataPath;
+  private ProcessProvider processProvider;
+
+  @VisibleForTesting
+  SecureConnectProvider(ProcessProvider processProvider, String metadataPath) {
+    this.processProvider = processProvider;
+    this.metadataPath = metadataPath;
+  }
+
+  public SecureConnectProvider() {
+    this(new DefaultProcessProvider(), DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+  }
+
+  /** The mutual TLS key store created with the default client certificate on device. */
+  @Override
+  public KeyStore getKeyStore() throws IOException {
+    try (InputStream stream = new FileInputStream(metadataPath)) {
+      return getKeyStore(stream, processProvider);
+    } catch (InterruptedException e) {
+      throw new IOException("Interrupted executing certificate provider command", e);
+    } catch (GeneralSecurityException e) {
+      throw new CertificateSourceUnavailableException(
+          "SecureConnect encountered GeneralSecurityException:", e);
+    } catch (FileNotFoundException exception) {
+      // If the metadata file doesn't exist, then there is no key store, so we will throw sentinel
+      // error
+      throw new CertificateSourceUnavailableException("SecureConnect metadata does not exist.");
+    }
+  }
+
+  @VisibleForTesting
+  static KeyStore getKeyStore(InputStream metadata, ProcessProvider processProvider)
+      throws IOException, InterruptedException, GeneralSecurityException {
+    Process process = processProvider.createProcess(metadata);
+
+    // Run the command and timeout after 1000 milliseconds.
+    int exitCode = runCertificateProviderCommand(process, 1000);
+    if (exitCode != 0) {
+      throw new IOException("Cert provider command failed with exit code: " + exitCode);
+    }
+
+    // Create mTLS key store with the input certificates from shell command.
+    return SecurityUtils.createMtlsKeyStore(process.getInputStream());
+  }
+
+  @VisibleForTesting
+  static ImmutableList<String> extractCertificateProviderCommand(InputStream contextAwareMetadata)
+      throws IOException {
+    JsonParser parser = new GsonFactory().createJsonParser(contextAwareMetadata);
+    ContextAwareMetadataJson json = parser.parse(ContextAwareMetadataJson.class);
+    return json.getCommands();
+  }
+
+  @VisibleForTesting
+  static int runCertificateProviderCommand(Process commandProcess, long timeoutMilliseconds)
+      throws IOException, InterruptedException {
+    long startTime = System.currentTimeMillis();
+    long remainTime = timeoutMilliseconds;
+
+    // In the while loop, keep checking if the process is terminated every 100 milliseconds
+    // until timeout is reached or process is terminated. In getKeyStore we set timeout to
+    // 1000 milliseconds, so 100 millisecond is a good number for the sleep.
+    while (remainTime > 0) {
+      Thread.sleep(Math.min(remainTime + 1, 100));
+      remainTime -= System.currentTimeMillis() - startTime;
```

Comment:
Instead of manual timeout logic implementation, for Java 8+, `Process.waitFor(long timeout, TimeUnit unit)` is the standard and much cleaner way to wait for a process with a timeout. 
If check "every 100 milliseconds" itself is not a requirement, this would simplify logic here, to something along the lines of
```
static int runCertificateProviderCommand(Process commandProcess, long timeoutMilliseconds)
    throws IOException, InterruptedException {
  boolean terminated = commandProcess.waitFor(timeoutMilliseconds, TimeUnit.MILLISECONDS);
  if (!terminated) {
    commandProcess.destroy();
    throw new IOException("Cert provider command timed out");
  }
  return commandProcess.exitValue();
}
```

Reply:
Done~